File: src/main/java/com/aiworkshop/aiworkshop/service/ArticleService.java
package com.aiworkshop.aiworkshop.service;

import java.util.List;

import org.springframework.stereotype.Service;

import com.aiworkshop.aiworkshop.dto.ArticleDto;
import com.aiworkshop.aiworkshop.exception.ResourceNotFoundException;
import com.aiworkshop.aiworkshop.mapper.ArticleMapper;
import com.aiworkshop.aiworkshop.repository.ArticleRepository;
import com.aiworkshop.aiworkshop.repository.FileRepository;
import com.aiworkshop.aiworkshop.repository.UserRepository;

import lombok.RequiredArgsConstructor;

@Service
@RequiredArgsConstructor
public class ArticleService {

    private final ArticleMapper mapper;
    private final FileService fileService;
    private final ArticleRepository repository;
    private final FileRepository fileRepository;
    private final UserRepository userRepository;

    public List<ArticleDto> getAll() {
        return repository
                .findAll()
                .stream()
                .map(mapper::toDto)
                .toList();
    }

    public ArticleDto getById(Long id) {
        final var article = repository
                .findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Article not found"));

        return mapper.toDto(article);
    }

    public ArticleDto create(ArticleDto dto) {
        final var username = dto.getUsername();
        final var imageIds = dto.getImageIds();
        final var contentId = dto.getContentId();
        final var isEmpty = imageIds.isEmpty();

        if (isEmpty) {
            throw new ResourceNotFoundException("Image not found");
        }

        final var user = userRepository
                .findByUsername(username)
                .orElseThrow(() -> new ResourceNotFoundException("User not found"));
        final var images = fileRepository.findAllById(imageIds);
        final var article = mapper.toEntity(dto, user, images);
        final var savedArticle = repository.save(article);
        final var id = savedArticle.getId();
        final var content = fileRepository
                .findById(contentId)
                .orElseThrow(() -> new ResourceNotFoundException("Content not found"));

        content.setArticle(savedArticle);
        fileRepository.save(content);
        fileService.updateAll(id, imageIds);

        return mapper.toDto(savedArticle);
    }

    public ArticleDto update(ArticleDto dto) {
        final var id = dto.getId();
        final var username = dto.getUsername();
        final var user = userRepository
                .findByUsername(username)
                .orElseThrow(() -> new ResourceNotFoundException("User not found"));
        final var article = repository
                .findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Article not found"));
        final var images = fileRepository.findAllById(dto.getImageIds());

        mapper.update(dto, images, user, article);

        final var savedArticle = repository.save(article);

        return mapper.toDto(savedArticle);
    }

    public void delete(Long id) {
        repository.deleteById(id);
    }

}

File: service/article_service.py
import json
from redis import Redis
from mapper.article_mapper import ArticleMapper
from model.user import User
from repository.article_repository import ArticleRepository
from exception.article_not_found_exception import ArticleNotFoundException
from dto.article_dto import ArticleDto, CreateArticleDto, UpdateArticleDto
from repository.user_repository import UserRepository


class ArticleService:
    def __init__(
        self, repository: ArticleRepository, user_repository: UserRepository, mapper: ArticleMapper, redis_client: Redis
    ):
        self.mapper = mapper
        self.repository = repository
        self.redis_client = redis_client
        self.user_repository = user_repository

    async def get_all(self) -> list[ArticleDto]:
        cache_key = "articles:all"
        cached = self.redis_client.get(cache_key)

        if cached:
            return [ArticleDto.model_validate_json(article) for article in json.loads(cached)]

        articles = await self.repository.get_all()
        result = [self.mapper.to_dto(dto_model=ArticleDto, orm_model=article) for article in articles]
        serialized = json.dumps([article.model_dump_json() for article in result])

        self.redis_client.set(cache_key, serialized, ex=86400)

        return result

    async def get_by_id(self, id: int) -> ArticleDto:
        cache_key = f"article:{id}"
        cached = self.redis_client.get(cache_key)

        if cached:
            return ArticleDto.model_validate_json(cached)

        article = await self.repository.get_by_id(id)

        if not (article):
            raise ArticleNotFoundException()

        user = await self.user_repository.get_user_by_id(article.user_id)
        dto = self.mapper.to_dto(dto_model=ArticleDto, orm_model=article)
        dto.username = user.username if user else None
        self.redis_client.set(cache_key, dto.model_dump_json(), ex=86400)
        return dto

    async def create(self, dto: CreateArticleDto, user: User) -> ArticleDto:
        article_dict = self.mapper.to_dict(dto)
        article_dict["user_id"] = user.id
        created_article = await self.repository.create(article_dict)

        self.redis_client.delete("articles:all")
        return self.mapper.to_dto(orm_model=created_article, dto_model=ArticleDto)

    async def update(self, id: int, dto: UpdateArticleDto, user: User) -> None:
        article_dict = self.mapper.to_dict(dto)
        article_dict["user_id"] = user.id

        await self.repository.update(id=id, article_dict=article_dict)
        self.redis_client.delete(f"article:{id}")
        self.redis_client.delete("articles:all")

    async def delete(self, id: int) -> None:
        await self.repository.delete(id)
        self.redis_client.delete(f"article:{id}")
        self.redis_client.delete("articles:all")
